File: PluginMessagingUtils/src/main/java/org/openredstone/messages/ReportMessage.java
package org.openredstone.messages;

import net.md_5.bungee.api.ChatColor;
import org.openredstone.types.Report;

import java.util.Arrays;

public class ReportMessage extends Message {

    Report report;

    public ReportMessage(Report report, String[] arguments) {
        super(arguments);
        this.report = report;
    }

    public ReportMessage(String serializedMessage) throws Exception {
        String[] raw = serializedMessage.split(":");

        if (raw.length < 1) {
            throw new Exception("Not enough arguments provided in serialized message.");
        }

        if (!isValidReport(raw[0])) {
            throw new IllegalArgumentException("Invalid action: " + raw[0]);
        }

        this.report = Report.valueOf(raw[0]);
        this.arguments = Arrays.copyOfRange(raw, 2, raw.length);

    }

    @Override
    public String getSerializedMessage() {
        String reportString = report.name();
        return reportString + ":" + super.getSerializedMessage();
    }

    public ChatColor getColor() {
        switch (report) {
            case INFO:
                return ChatColor.WHITE;
            case WARNING:
                return ChatColor.YELLOW;
            case ERROR:
                return ChatColor.RED;
        }
        return ChatColor.RESET;
    }

    private static boolean isValidReport(String report) {
        try {
            Report.valueOf(report);
            return true;
        } catch (IllegalArgumentException e) {
            return false;
        }
    }

    public Report getReport() {
        return report;
    }

}

File: plugin-messaging/src/main/kotlin/org/openredstone/messages/ReportMessage.kt
package org.openredstone.messages

import net.md_5.bungee.api.ChatColor
import org.openredstone.types.Report

class ReportMessage {
    var arguments: Array<String>
    var report: Report

    constructor(report: Report, arguments: Array<String>) {
        this.arguments = arguments
        this.report = report
    }

    constructor(serializedMessage: String) {
        val raw: Array<String> = serializedMessage.split(":").toTypedArray()
        if (raw.isEmpty()) {
            throw Exception("Not enough arguments provided in serialized message.")
        }
        require(isValidReport(raw[0])) { "Invalid action: " + raw[0] }
        report = Report.valueOf(raw[0])
        arguments = raw.copyOfRange(2, raw.size)
    }

    fun getSerializedMessage(): String =
        report.name + ":" + arguments.joinToString(":")

    val color: ChatColor
        get()
        = when (report) {
            Report.INFO -> ChatColor.WHITE
            Report.WARNING -> ChatColor.YELLOW
            Report.ERROR -> ChatColor.RED
        }

    companion object {
        private fun isValidReport(report: String?): Boolean {
            return try {
                Report.valueOf(report!!)
                true
            } catch (e: IllegalArgumentException) {
                false
            }
        }
    }
}
